engine/apps/slack/migrations/0009_rename__slack_team_identity_slackmessage_slack_team_identity_and_more.py

```diff
@@ -0,0 +1,63 @@
+# Generated by Django 4.2.16 on 2024-12-04 21:17
+
+import logging
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django_migration_linter as linter
+
+logger = logging.getLogger(__name__)
+
+
+def drop_orphaned_slack_messages(apps, schema_editor):
+    """
+    At this point, in the 0007_migrate_slackmessage_channel_id migration, we have already migrated the old
+    SlackMessage._channel_id (char field) to the new SlackMessage.channel_id (FK field).
+
+    This script is needed to identity, and drop, SlackMessage records that are "orphaned". This would've occured in two
+    cases:
+    - the organization switched their Slack team identity. We have historically never done clean-up of the old SlackMessage records
+    - a Slack channel was deleted. Previously, we did not have a FK constraint on the SlackMessage._channel_id
+    char field, so there was no CASCADE delete behaviour (now there is).
+
+    In both of these scenarios, the SlackMessage records would be "orphaned" - they would have a non-null channel_id
+    foreign key relationship, which prevents us from setting NOT NULL constraint on the channel_id field.
+    """
+    SlackMessage = apps.get_model("slack", "SlackMessage")
+
+    logger.info("Starting migration to populate the 'channel' field.")
+
+    slack_messages = SlackMessage.objects.filter(channel_id__isnull=True)
+
+    total_messages = slack_messages.count()
+    if total_messages == 0:
+        logger.info("No orphaned SlackMessages need to be dropped.")
+        return
+
+    logger.info(f"Found {total_messages} orphaned SlackMessages to drop.")
+
+    slack_messages.delete()
+
+    logger.info("Finished migration to drop orphaned SlackMessages.")
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('slack', '0008_remove_slackmessage_active_update_task_id_state'),
+    ]
+
+    operations = [
+        linter.IgnoreMigration(),
+        migrations.RenameField(
+            model_name='slackmessage',
+            old_name='_slack_team_identity',
+            new_name='slack_team_identity',
+        ),
+        migrations.RunPython(drop_orphaned_slack_messages, migrations.RunPython.noop),
```

Comment:
TODO: before this step, add another `migrations.RunPython` step to fill in `null` `SlackMessage.slack_team_identity` values.

There was a brief period of time where we (incorrectly) switched over to _just_ filling in `slack_team_identity` when creating `SlackMessage` records. No problem, we have access to this information 🙂